engine/api/workflowtemplate/import.go

```diff
@@ -0,0 +1,140 @@
+package workflowtemplate
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/go-gorp/gorp"
+	"github.com/ovh/cds/engine/api/event"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/exportentities"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// Push creates or updates a workflow template from a tar.
+func Push(db gorp.SqlExecutor, u *sdk.User, tr *tar.Reader) ([]sdk.Message, *sdk.WorkflowTemplate, error) {
+	// extract template data from tar
+	var apps, pips, envs [][]byte
+	var wkf []byte
+	var tmpl exportentities.Template
+
+	mError := new(sdk.MultiError)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, sdk.NewError(sdk.ErrWrongRequest, sdk.WrapError(err, "Unable to read tar file"))
+		}
+
+		buff := new(bytes.Buffer)
+		if _, err := io.Copy(buff, tr); err != nil {
+			return nil, nil, sdk.NewError(sdk.ErrWrongRequest, sdk.WrapError(err, "Unable to read tar file"))
+		}
+
+		var templateFileName string
+		b := buff.Bytes()
+		switch {
+		case strings.Contains(hdr.Name, ".application."):
+			apps = append(apps, b)
+		case strings.Contains(hdr.Name, ".pipeline."):
+			pips = append(pips, b)
+		case strings.Contains(hdr.Name, ".environment."):
+			envs = append(envs, b)
+		case hdr.Name == "workflow.yml":
+			// if a workflow was already found, it's a mistake
+			if len(wkf) != 0 {
+				mError.Append(fmt.Errorf("Two workflow files found"))
+				break
+			}
+			wkf = b
+		default:
+			// if a template was already found, it's a mistake
+			if templateFileName != "" {
+				mError.Append(fmt.Errorf("Two template files found: %s and %s", templateFileName, hdr.Name))
+				break
+			}
+			if err := yaml.Unmarshal(b, &tmpl); err != nil {
+				mError.Append(sdk.WrapError(err, "Unable to unmarshal template %s", hdr.Name))
+				continue
+			}
+			templateFileName = hdr.Name
+		}
+	}
+
+	if !mError.IsEmpty() {
+		return nil, nil, sdk.NewError(sdk.ErrWorkflowInvalid, mError)
+	}
+
+	// init workflow template struct from data
+	wt := tmpl.GetTemplate()
+	wt.Value = base64.StdEncoding.EncodeToString(wkf)
+	for _, p := range pips {
```

Comment:
I think you can initialise your slice with right length/capacity

Reply:
done